paho/subscriber.go

```diff
@@ -0,0 +1,203 @@
+	// in as parameters to `Run` and never stored. This way we can ensure that nothing changes 
+	// underneath us.
+
+	// Note 3: Any objects that might normally need thread protection (like the `ackMap` object) are 
+	// stored as stack parameters in this function and never passed around. Because nobody else is
+	// using them, we don't need to worry about protecting them. 
+	// lock them.
+
+	// Keep it simple for now. This could use a PacketID allocater later. 
+	ackMap := make(map[uint16](*InternalSend))
+	var nextPacketID uint16 = 1
+
+	defer s.Cleanup()
+
+	s.debug.Println("SubUnsubHandler running")
+
+	for {
+		select {
+		case <- stop:
+			s.debug.Println("SubUnsubHandler stopped")
+			// We don't have to worry about cleaning up ackMap.  Any subscribe or unsubscribe 
+			// operations have been handled by the Subscribe and Unsubscribe calls.
+
+			// Since were returning, the `defer s.Cleanup()` code above ensures that Subscribe
+			// and Unsubscribe operations get stopped (via `close s.internalStop`)
+			return
+		case is := <- s.internalSend:
+			// This is a message from either `Subscribe` or `Unsubscribe`. This code needs to be
+			// here because 1. we are the only function allowed to push to `send` and 2. we are the
+			// only function alloed to modify `ackMap`.
+			packetID := nextPacketID
+			// TODO what happens on overflow. Does it wrap around again?
+			nextPacketID++
+			ackMap[packetID] = is
+
+			is.packet.SetPacketID(packetID)
+
+			// Wrap this in a goroutine since we don't care when this completes and we'll catch
+			// failures some other way (probably via timeout or failure reading from the socket).
+			go func() {
+				s.debug.Printf("SubUnsubHandler sending %v packetId %v", is.packet.PacketType(), packetID)
+				send <- is.packet
+			}()
+		case pid := <- s.internalCancel:
+			// This is anoher message from `Subscribe` or `Unsubscribe` that tells when a sub or unsub
+			// operation has been cancelled.
+			delete(ackMap, pid)
+		case a := <- s.Receive:
+			// This is a message from the socket.  It should be either a `packets.Suback` or 
+			// `packets.Unsuback` object since they are the only ones we need to worry about.
+			s.debug.Printf("SubUnsubHandler received %v packetId %v", a.PacketType(), a.PacketID())
+			is, ok := ackMap[a.PacketID()]
+			if ok {
+				delete(ackMap, a.PacketID())
+				go func() {
+					// Trigger the completion of the `Subscribe` or `Unsubscribe` call.  Do this in a
+					// goroutine because we don't need to block this loop waiting for any receivers to 
+					// become ready.
+					is.ackReceived <- a
+				}()
+			}
+		}
+	}
+}
+
+// Send a subscribe packet and wait for the matching suback to return. This is safe to call from
+// a goroutine.
+func (s *SubUnsubHandler) Subscribe(ctx context.Context, subscribe *packets.Subscribe) (*packets.Suback, error) {
+	ackReceived := make(chan *packets.ControlPacket)
+	defer close(ackReceived)
```

Comment:
I think it's considered bad for a reader to close a channel, I think the writer is supposed to close it

Reply:
Writing to a closed channel causes a panic